File: frontend/app/__tests__/api-contract.test.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";

import type {
  ApiClub,
  ApiEvent,
  ApiMyEvent,
  AuthResponse,
  ClubAdmin,
  ClubAdminRequest,
  ClubAuditLog,
  ClubInvitation,
  Interest,
  ManagedClub,
  NotificationPreferences,
  OwnershipTransfer,
  ProfileSocialLinks,
  User,
  UserProfile,
} from "@/app/types";

/**
 * The frontend half of the API contract.
 *
 * `app/types/index.ts` is a hand-written mirror of the backend's records —
 * there is no OpenAPI generation and no shared package. So renaming a field in
 * `EventDTO` used to break this app with nothing failing anywhere: the backend
 * suite passed, this suite passed, and the defect surfaced in a browser.
 * Neither side can catch that alone, because neither knows the other's types.
 *
 * Both sides now assert against one file, `contracts/api-dto-fields.json`.
 * `ApiContractTest.java` proves the backend serialises exactly those names;
 * this proves the interfaces below declare exactly those names. Agreeing with
 * the same file is what makes them agree with each other.
 *
 * There are two layers here, and they catch different mistakes:
 *
 *   - `Record<keyof T, true>` is a COMPILE-TIME check. Remove a field from the
 *     interface and the object below has an excess property; add one and the
 *     object is missing a key. Either way `npm run type-check` fails, naming
 *     the field, before any test runs.
 *   - The assertions are a RUNTIME check against the backend's list. This is
 *     the one that catches the case that matters: the interface and the record
 *     drifting apart.
 */

const CONTRACT: Record<string, string[]> = JSON.parse(
  readFileSync(
    resolve(__dirname, "..", "..", "..", "contracts", "api-dto-fields.json"),
    "utf8",
  ),
);

// Keying on `keyof T` is the whole point — these are not free-standing lists of
// strings, they are exhaustive over the interface. Do not replace them with
// `string[]`, which would compile happily while the interface changed.
const eventFields: Record<keyof ApiEvent, true> = {
  id: true,
  title: true,
  description: true,
  dateTime: true,
  createdAt: true,
  location: true,
  price: true,
  organizerId: true,
  organizerName: true,
  followers: true,
  images: true,
  promoted: true,
  capacity: true,
  registered: true,
  categories: true,
};

const clubFields: Record<keyof ApiClub, true> = {
  id: true,
  name: true,
  description: true,
  followers: true,
  logo: true,
  socialLinks: true,
  featured: true,
  images: true,
  createdAt: true,
};

const userFields: Record<keyof User, true> = {
  id: true,
  name: true,
  email: true,
  roles: true,
  createdAt: true,
  emailVerified: true,
  authProvider: true,
};

const myEventFields: Record<keyof ApiMyEvent, true> = {
  event: true,
  going: true,
  saved: true,
};

const authResponseFields: Record<keyof AuthResponse, true> = {
  token: true,
  user: true,
};

const clubAdminRequestFields: Record<keyof ClubAdminRequest, true> = {
  id: true,
  userId: true,
  userName: true,
  userEmail: true,
  clubId: true,
  clubName: true,
  message: true,
  status: true,
  requestedAt: true,
  reviewedAt: true,
};

const clubAdminFields: Record<keyof ClubAdmin, true> = {
  assignmentId: true,
  userId: true,
  userName: true,
  userEmail: true,
  invitedEmail: true,
  role: true,
  status: true,
  createdAt: true,
  activatedAt: true,
};

const clubInvitationFields: Record<keyof ClubInvitation, true> = {
  invitationId: true,
  clubId: true,
  clubName: true,
  clubLogo: true,
  role: true,
  invitedByName: true,
  invitedAt: true,
};

const ownershipTransferFields: Record<keyof OwnershipTransfer, true> = {
  transferId: true,
  clubId: true,
  clubName: true,
  clubLogo: true,
  fromUserId: true,
  fromUserName: true,
  toUserId: true,
  toUserName: true,
  outgoingBecomes: true,
  status: true,
  createdAt: true,
};

const clubAuditLogFields: Record<keyof ClubAuditLog, true> = {
  id: true,
  action: true,
  entityType: true,
  entityId: true,
  actorUserId: true,
  actorName: true,
  metadata: true,
  createdAt: true,
};

const managedClubFields: Record<keyof ManagedClub, true> = {
  clubId: true,
  clubName: true,
  logo: true,
  followers: true,
  role: true,
  officialEmail: true,
  officialEmailVerified: true,
};

const userProfileFields: Record<keyof UserProfile, true> = {
  bio: true,
  faculty: true,
  degree: true,
  subjects: true,
  socialLinks: true,
  interests: true,
  showInterests: true,
  showSocialLinks: true,
};

// Pinned separately because the contract records a nested object as one field
// name: `userProfileFields` above contributes `socialLinks` and says nothing
// about what is inside it. Without an entry of its own, renaming `instagram` on
// either side would break the app with both suites still green.
const profileSocialLinksFields: Record<keyof ProfileSocialLinks, true> = {
  instagram: true,
  facebook: true,
  linkedin: true,
};

const notificationPreferencesFields: Record<keyof NotificationPreferences, true> = {
  eventReminders: true,
  clubAnnouncements: true,
  weeklyDigest: true,
  newFollowerEvents: true,
  productNews: true,
};

const interestFields: Record<keyof Interest, true> = {
  slug: true,
  label: true,
  category: true,
};

/** Backend DTO name → the TypeScript interface mirroring it. */
const MIRRORS: Record<string, Record<string, true>> = {
  EventDTO: eventFields,
  ClubDTO: clubFields,
  UserDTO: userFields,
  MyEventDTO: myEventFields,
  AuthenticationResponse: authResponseFields,
  ClubAdminRequestDTO: clubAdminRequestFields,
  ClubAdminDTO: clubAdminFields,
  ClubInvitationDTO: clubInvitationFields,
  OwnershipTransferDTO: ownershipTransferFields,
  ClubAuditLogDTO: clubAuditLogFields,
  ManagedClubDTO: managedClubFields,
  UserProfileDTO: userProfileFields,
  ProfileSocialLinksDTO: profileSocialLinksFields,
  NotificationPreferencesDTO: notificationPreferencesFields,
  InterestDTO: interestFields,
};

describe("API contract", () => {
  it.each(Object.keys(MIRRORS))(
    "%s — the TypeScript mirror declares exactly the agreed fields",
    (dto) => {
      expect(CONTRACT[dto]).toBeDefined();
      expect(Object.keys(MIRRORS[dto]).sort()).toEqual([...CONTRACT[dto]].sort());
    },
  );

  it("mirrors every type the contract names", () => {
    // Without this, adding a DTO to the contract without a TypeScript mirror
    // passes silently — the loop above only checks what it already knows about.
    // Keys beginning with $ are documentation, not types.
    const documented = Object.keys(CONTRACT)
      .filter((key) => !key.startsWith("$"))
      .sort();

    expect(documented).toEqual(Object.keys(MIRRORS).sort());
  });
});

File: backend/src/test/java/com/campusvibe/contract/ApiContractTest.java
package com.campusvibe.contract;

import com.campusvibe.auth.AuthenticationResponse;
import com.campusvibe.club.ClubDTO;
import com.campusvibe.clubadmin.ClubAdminDTO;
import com.campusvibe.clubadmin.ClubAuditLogDTO;
import com.campusvibe.clubadmin.ClubInvitationDTO;
import com.campusvibe.clubadmin.OwnershipTransferDTO;
import com.campusvibe.clubadmin.ClubAdminRequestDTO;
import com.campusvibe.clubadmin.ManagedClubDTO;
import com.campusvibe.event.EventDTO;
import com.campusvibe.user.MyEventDTO;
import com.campusvibe.user.UserDTO;
import com.campusvibe.taxonomy.ClubCategoryDTO;
import com.campusvibe.taxonomy.EventFormatDTO;
import com.campusvibe.taxonomy.InterestDTO;
import com.campusvibe.user.profile.NotificationPreferencesDTO;
import com.campusvibe.user.profile.ProfileSocialLinksDTO;
import com.campusvibe.user.profile.UserProfileDTO;
import com.fasterxml.jackson.databind.BeanDescription;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.introspect.BeanPropertyDefinition;
import org.assertj.core.api.SoftAssertions;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * The backend half of the API contract.
 *
 * <p>{@code frontend/app/types/index.ts} is a hand-written mirror of these
 * records. There is no OpenAPI generation and no shared package, so renaming a
 * field here used to break the frontend with nothing failing anywhere: this
 * suite passed, the frontend suite passed, and the defect surfaced in a browser.
 * Neither side can catch that alone, because neither knows the other's types.
 *
 * <p>Both sides now assert against one file,
 * {@code contracts/api-dto-fields.json}. This test proves the backend still
 * serialises exactly those names; its counterpart,
 * {@code frontend/app/__tests__/api-contract.test.ts}, proves the TypeScript
 * interfaces declare exactly those names. Agreeing with the same file is what
 * makes them agree with each other.
 *
 * <p>Deliberately a plain unit test — no Spring context. It reads type metadata,
 * not behaviour, so it costs milliseconds and fails on the commit that breaks
 * the contract rather than on the one that deploys it.
 */
class ApiContractTest {

    /**
     * Every DTO the frontend mirrors. A record that no TypeScript interface
     * copies does not belong here — the contract is for types that cross the
     * wire into hand-written code, not for every DTO in the codebase.
     */
    private static final Map<String, Class<?>> CONTRACTED = new LinkedHashMap<>();

    static {
        CONTRACTED.put("EventDTO", EventDTO.class);
        CONTRACTED.put("ClubDTO", ClubDTO.class);
        CONTRACTED.put("UserDTO", UserDTO.class);
        CONTRACTED.put("MyEventDTO", MyEventDTO.class);
        CONTRACTED.put("AuthenticationResponse", AuthenticationResponse.class);
        CONTRACTED.put("ClubAdminRequestDTO", ClubAdminRequestDTO.class);
        CONTRACTED.put("ClubAdminDTO", ClubAdminDTO.class);
        CONTRACTED.put("ClubInvitationDTO", ClubInvitationDTO.class);
        CONTRACTED.put("OwnershipTransferDTO", OwnershipTransferDTO.class);
        CONTRACTED.put("ClubAuditLogDTO", ClubAuditLogDTO.class);
        CONTRACTED.put("ManagedClubDTO", ManagedClubDTO.class);
        CONTRACTED.put("UserProfileDTO", UserProfileDTO.class);
        // Contracted in its own right because the contract records a
        // nested object as a single field name -- UserProfileDTO
        // contributes "socialLinks" and nothing about what is inside it,
        // so without this entry these three names cross the wire
        // unchecked.
        CONTRACTED.put("ProfileSocialLinksDTO", ProfileSocialLinksDTO.class);
        CONTRACTED.put("NotificationPreferencesDTO", NotificationPreferencesDTO.class);
        CONTRACTED.put("InterestDTO", InterestDTO.class);
        CONTRACTED.put("ClubCategoryDTO", ClubCategoryDTO.class);
        CONTRACTED.put("EventFormatDTO", EventFormatDTO.class);
    }

    private final ObjectMapper mapper = new ObjectMapper();

    @Test
    void everyContractedDtoSerialisesExactlyTheAgreedFieldNames() throws IOException {
        JsonNode contract = mapper.readTree(contractFile().toFile());

        // Reported together rather than one per run: a rename usually moves two
        // or three fields at once, and finding out about the second one only
        // after fixing the first is a slow way to learn what changed.
        SoftAssertions softly = new SoftAssertions();

        CONTRACTED.forEach((name, type) -> softly.assertThat(serialisedFieldNames(type))
                .as("%s — update contracts/api-dto-fields.json and the TypeScript "
                        + "mirror in the same commit as the record", name)
                .isEqualTo(agreedFieldNames(contract, name)));

        softly.assertAll();
    }

    @Test
    void theContractDoesNotNameATypeNobodyChecks() throws IOException {
        // Without this, deleting an entry from CONTRACTED silently stops
        // checking that DTO while leaving the contract file looking complete.
        JsonNode contract = mapper.readTree(contractFile().toFile());

        Set<String> documented = new TreeSet<>();
        contract.fieldNames().forEachRemaining(field -> {
            if (!field.startsWith("$")) {
                documented.add(field);
            }
        });

        assertThat(documented).isEqualTo(new TreeSet<>(CONTRACTED.keySet()));
    }

    /**
     * The names Jackson will actually put on the wire.
     *
     * <p>Introspection rather than serialising an instance: it needs no
     * populated object, and it reports the field even when its value would be
     * null. A test that built instances would silently stop checking any field
     * left unset.
     */
    private Set<String> serialisedFieldNames(Class<?> type) {
        BeanDescription description = mapper.getSerializationConfig()
                .introspect(mapper.constructType(type));
        Set<String> names = new TreeSet<>();
        for (BeanPropertyDefinition property : description.findProperties()) {
            names.add(property.getName());
        }
        return names;
    }

    private Set<String> agreedFieldNames(JsonNode contract, String dto) {
        JsonNode names = contract.get(dto);
        assertThat(names)
                .as("contracts/api-dto-fields.json has no entry for %s", dto)
                .isNotNull();

        Set<String> agreed = new TreeSet<>();
        names.forEach(name -> agreed.add(name.asText()));
        return agreed;
    }

    /**
     * Surefire runs with the working directory at {@code backend/}, so the
     * contract sits one level up. Resolved to an absolute path before the
     * existence check purely so the failure message says where it looked.
     */
    private Path contractFile() {
        Path path = Path.of("..", "contracts", "api-dto-fields.json").toAbsolutePath().normalize();
        assertThat(Files.exists(path))
                .as("API contract not found at %s — it is shared with the frontend "
                        + "and must not be moved without updating both tests", path)
                .isTrue();
        return path;
    }
}
